pkg/apis/topology/v1alpha1/types_partitionset.go

```diff
@@ -58,16 +60,47 @@ type PartitionSetSpec struct {
 
 	// +optional
 
-	// selector (optional) is a label selector that filters shard targets.
-	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	// MatchExpressions (optional) specifies filtering for shard targets.
+	// MatchExpressions are used here instead of Selector to avoid possible collisions with
+	// the MatchLabels generated for the Partition according to the dimensions in the PartitionSet.
+	// This does not limit the filtering as any MatchLabel can also be expressed as
+	// a MatchExpression.
+	MatchExpressions []metav1.LabelSelectorRequirement `json:"matchExpressions,omitempty"`
```

Comment:
Would it make sense to make `PartitionSetSpec.Dimensions` and/or `PartitionSetSpec.matchExpressions/PartitionSetSpec.selector` immutable ?

Reply:
As I see it this is not the likely flow.  This does not take in considerations that Partitions need to be colocated with the APIExportEndpointSlices that consume them. Here is how I see the flow:
- A service provider update a PartitionSet
- The set of Partitions bound to the PartitionSet is updated accordingly in the PartitionSet workspace
- Each of this Partition is copied into the workspace where the intended APIExportEndpointSlice is located. Each APIExportEndpointSlice is intended for an operator/controller manager instance. It makes URLs for a set of shards identified by the Partition available to it.
- The service provider can choose at any convenient time to patch the APIExportEndpointSlices with the new Partitions
There is no modification of the service and no disruption until the APIExportEndpointSlices are patched.

The steps above could be automated but it is another discussion.

Reply:
I'm not sure if I read the code correctly but it seems that we are deleting old partitions potentially being used by APIExportEndpointSlice which seems to be disruptive.

xref: https://github.com/kcp-dev/kcp/blob/c991948b1e6f7b14152b0a2714930b1b1c42b4b5/pkg/reconciler/topology/partitionset_reconcile.go#L123